Support n-ary monotonic functions in discover_new_orderings

gokselk opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

gokselk committed:
Support n-ary monotonic functions in `discover_new_orderings`

gokselk committed:
Add tests for n-ary monotonic functions in `discover_new_orderings`

berkaysynnada commented:
I'm sorry for the delayed response @gokselk. I had to deal with some urgent maintenance issues. The refactor looks promising overall, but I noticed some issues in the tests that indicate potential bugs.

In your first test, the table is ordered as [a+b, a, b], and you add the equality c = a + b. This results in new orderings of [a+b] and [a, b]. I believe this is incorrect—you can find a simple counter-example if this isn’t immediately clear. The initial orderings should remain unchanged (or potentially be updated to [c, a, b] but this is not our concern).

The correct test case should be as follows: start with an initial ordering of [c, a, b], then add the equality c = a + b. The final orderings should then be [c] and [a, b]. I tested this scenario with your implementation, and it works as expected (it does not work without your refactor). However, the current test should not pass, as the implementation seems to perform unintended simplifications. You need to identify and address this bug.

Once that’s fixed, it would be beneficial to add more scenarios. Consider using `ScalarFunctionExpr`'s as PhysicalExpr to represent different mathematical functions.

berkaysynnada commented:
You also need to resolve the conflicts coming from upstream to your branch.

gokselk committed:
Fix tests

gokselk commented:
> In your first test, the table is ordered as [a+b, a, b], and you add the equality c = a + b. This results in new orderings of [a+b] and [a, b]. I believe this is incorrect—you can find a simple counter-example if this isn’t immediately clear. The initial orderings should remain unchanged (or potentially be updated to [c, a, b] but this is not our concern).
> 
> The correct test case should be as follows: start with an initial ordering of [c, a, b], then add the equality c = a + b. The final orderings should then be [c] and [a, b]. I tested this scenario with your implementation, and it works as expected (it does not work without your refactor). However, the current test should not pass, as the implementation seems to perform unintended simplifications. You need to identify and address this bug.

Thanks for catching this. I agree the original test case was incorrect. I'm debugging the unintended simplification behavior with the ordering [a+b, a, b]. Will update once I identify and fix the issue.

For now, I've updated the test to use the correct case: [c, a, b] -> [c] and [a, b].

gokselk committed:
Fix non-monotonic test case

gokselk committed:
Fix unintended simplification

gokselk commented:
@berkaysynnada I've pushed a potential fix that avoids unintended simplification by skipping cases where the original ordering starts with the equivalent expression.

However, I'm not entirely confident about this approach and would appreciate your review on whether this is the right way to handle these cases.

gokselk committed:
Minor comment changes

gokselk committed:
Fix tests

berkaysynnada reviewed (commented):
I have some concern about the last test. It should behave in the same way with the first test, but results in a different state. To clear this up and discussing some ways to simplify implementation, do you want to set a meeting ASAP? We can also talk about the last task.

However, you can open this also to the upstream. It is clear that it brings an improvement, and we can continue the discussion there.

gokselk committed:
Add `preserves_lex_ordering` field

gokselk committed:
Use `preserves_lex_ordering` on `discover_new_orderings()`

gokselk committed:
Add `output_ordering` and `output_preserves_lex_ordering` implementations for `ConcatFunc`

gokselk committed:
Update tests

gokselk commented:
@berkaysynnada 

I've implemented the changes we discussed:
- Added `preserves_lex_ordering` functionality
- Updated test cases to use `concat` instead of addition.

Please review when you have a chance. Let me know if anything else needs to be addressed.

berkaysynnada commented on datafusion/functions/src/string/concat.rs:
```diff
@@ -261,6 +262,23 @@ impl ScalarUDFImpl for ConcatFunc {
         Some(get_concat_doc())
     }
+
+    fn output_ordering(&self, inputs: &[ExprProperties]) -> Result<SortProperties> {
```
Move that logic to fn output_preserves_lex_ordering trait impl, and remove this override

gokselk replied:
Resolved in 8949021.

berkaysynnada commented on datafusion/functions/src/string/concat.rs:
```diff
@@ -261,6 +262,23 @@ impl ScalarUDFImpl for ConcatFunc {
+            .unwrap_or(SortProperties::Unordered))
+    }
+
+    fn output_preserves_lex_ordering(&self, inputs: &[ExprProperties]) -> Result<bool> {
```
Why don't you directly return true here ?

gokselk replied:
Resolved in 8949021.

berkaysynnada commented on datafusion/expr/src/udf.rs:
```diff
@@ -646,6 +653,11 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
         Ok(SortProperties::Unordered)
```
Instead of directly returning Unordered, lets add an if check:
```rust
if self.output_preserves_lex_ordering() {
    // now look the input orders, and if they are all going to the same direction (all asc or dsc), the output order will be that direction 
    ...
    } else {
        Unordered
    }
   ```

gokselk replied:
Resolved in 8949021.

berkaysynnada commented on datafusion/physical-expr/Cargo.toml:
```diff
@@ -47,6 +47,7 @@ chrono = { workspace = true }
 datafusion-common = { workspace = true, default-features = true }
 datafusion-expr = { workspace = true }
 datafusion-expr-common = { workspace = true }
+datafusion-functions = { workspace = true }
```
When you open this PR to the upstream, we should ask if adding this dependency is a problem. Can you take a note reminding that?

berkaysynnada reviewed (commented):
After you change the API as the way I suggest, we will not have to duplicate the order checking logic.

Did you also find a chance to investigate other functions? It will be better to have one more lex order preserver function

gokselk committed:
Move logic to UDF

gokselk committed:
Cargo fmt

gokselk commented:
> Did you also find a chance to investigate other functions? It will be better to have one more lex order preserver function

I'll investigate string manipulation functions (substr, case conversions, padding) and date formats as potential candidates for lex order preservation. Will update once I have concrete findings.

berkaysynnada commented on datafusion/expr/src/udf.rs:
```diff
@@ -649,13 +649,31 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     /// Whether the function preserves lexicographical ordering based on the input ordering
-    fn output_preserves_lex_ordering(&self, _inputs: &[ExprProperties]) -> Result<bool> {
-        Ok(false)
+    fn output_preserves_lex_ordering(&self, inputs: &[ExprProperties]) -> Result<bool> {
```
Isn't it just returning false?

gokselk replied:
> Isn't it just returning false?

No, the function doesn't simply return false. It returns true when there's at least one input and all inputs share identical sort properties. I've refactored the code to make this logic more readable.

berkaysynnada replied:
But the condition you've said is checked already at the output_ordering() part. This output_preserves_lex_ordering() is only for labeling the functions which preserves the lex ordering. You define it as false here, and only preserver functions override it as false. So, this API will not be open to outside, and the users still interact with the output_ordering() API. Am I missing smth?

gokselk replied:
Yes, makes sense now. Resolved in 542e1dc.

gokselk committed:
Refactor

gokselk committed:
Cargo fmt

gokselk committed:
Simply use false value on default implementation

gokselk committed:
Remove unnecessary import

berkaysynnada commented:
If the test pass successfully, let's open this to the upstream @gokselk

gokselk committed:
Clippy fix

gokselk committed:
Update Cargo.lock

gokselk committed:
Move dep to dev-dependencies

gokselk committed:
Rename output_preserves_lex_ordering to preserves_lex_ordering

berkaysynnada committed:
minor

ozankabak commented:
Merged upstream.
